src/Storages/StorageFile.cpp

```diff
@@ -910,6 +974,58 @@ namespace
     };
 }
 
+StorageFile::FileSource StorageFile::FileSource::parse(const String & source, const ContextPtr & context, std::optional<bool> allow_archive_path_syntax)
+{
+    String filename;
+    String path_to_archive;
+
+    if (!allow_archive_path_syntax)
+        allow_archive_path_syntax = context->getSettingsRef()[Setting::allow_archive_path_syntax];
+
+    if (*allow_archive_path_syntax)
+        std::tie(path_to_archive, filename) = splitToArchivePathAndPathInArchive(source);
+    else
+        filename = source;
+
+    FileSource res;
+    String user_files_path = context->getUserFilesPath();
+
+    if (!path_to_archive.empty())
+        res.archive_info = getArchiveInfo(path_to_archive, filename, user_files_path, context, res.total_bytes_to_read);
+    else
+        res.paths = getPathsList(filename, user_files_path, context, res.total_bytes_to_read);
+
+    res.with_globs = res.paths.size() > 1;
+
+    if (res.archive_info)
+    {
+        res.format_from_filenames = FormatFactory::instance().tryGetFormatFromFileName(res.archive_info->path_in_archive);
+    }
+    else
+    {
+        for (const String & path : res.paths)
+        {
+            auto single_file_format = FormatFactory::instance().tryGetFormatFromFileName(path);
+            if (!res.format_from_filenames)
+            {
+                res.format_from_filenames = single_file_format;
+            }
+            else if (res.format_from_filenames != single_file_format)
+            {
+                res.format_from_filenames = {};
+                break;
+            }
+        }
+    }
+
+    if (!res.paths.empty())
+        res.path_for_partitioned_write = res.paths.front();
+    else
+        res.path_for_partitioned_write = filename;
```

Comment:
`path_for_partitioned_write` can also be in the form of a method

Reply:
Yes, this calculation is simple, however it looks clear why `path_for_partitioned_write = res.paths.front()` (or `filename`) only when we parse `FileSource` from string. If we made this calculation later it would look strange. For this reason I'd prefer to keep this code as is.